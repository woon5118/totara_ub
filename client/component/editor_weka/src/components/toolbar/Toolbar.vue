<!--
  This file is part of Totara Enterprise Extensions.

  Copyright (C) 2020 onwards Totara Learning Solutions LTD

  Totara Enterprise Extensions is provided only to Totara
  Learning Solutions LTD's customers and partners, pursuant to
  the terms and conditions of a separate agreement with Totara
  Learning Solutions LTD or its affiliate.

  If you do not have an agreement with Totara Learning Solutions
  LTD, you may not access, use, modify, or distribute this software.
  Please contact [licensing@totaralearning.com] for more information.

  @author Simon Chester <simon.chester@totaralearning.com>
  @module editor_weka
-->

<template>
  <div
    class="tui-wekaToolbar"
    role="toolbar"
    :aria-label="$str('label_toolbar', 'editor_weka')"
  >
    <div class="tui-wekaToolbar__group">
      <Dropdown :separator="false">
        <template v-slot:trigger="{ toggle, isOpen }">
          <ToolbarButton
            class="tui-wekaToolbar__currentBlock tui-wekaToolbar__button"
            :text="activeBlockName"
            :aria-expanded="isOpen ? 'true' : 'false'"
            :aria-label="
              $str('format_as_blocktype_status', 'editor_weka', activeBlockName)
            "
            :disabled="!blockSelectorEnabled"
            caret
            @click="toggle()"
            @keydown.native="handleButtonKeydown"
            @focus.native="handleButtonFocus"
          />
        </template>
        <DropdownItem
          v-for="(item, i) in blockItems"
          :key="i"
          :selected="item.active"
          :disabled="!item.active && !item.enabled"
          @click="itemClick(item)"
        >
          {{ item.label.toString() }}
        </DropdownItem>
      </Dropdown>
    </div>
    <div
      v-for="(itemGroup, groupKey) in itemGroups"
      :key="groupKey"
      class="tui-wekaToolbar__group"
    >
      <template v-for="(item, i) in itemGroup">
        <!-- dropdown toolbar item -->
        <Dropdown v-if="item.children" :key="i" :separator="false">
          <template v-slot:trigger="{ toggle }">
            <ToolbarButtonIcon
              class="tui-wekaToolbar__button"
              :text="item.label.toString()"
              :active="item.active"
              :disabled="!item.enabled"
              @click="toggle"
              @keydown.native="handleButtonKeydown"
              @focus.native="handleButtonFocus"
            >
              <template slot="icon">
                <component :is="item.iconComponent" />
              </template>
            </ToolbarButtonIcon>
          </template>
          <DropdownItem
            v-for="(child, j) in item.children"
            :key="j"
            :disabled="!child.enabled"
            @click="itemClick(child)"
          >
            <component :is="child.iconComponent" />
            {{ child.label.toString() }}
          </DropdownItem>
        </Dropdown>
        <!-- popover toolbar component -->
        <Popover
          v-else-if="item.popover"
          :key="i"
          :triggers="['click']"
          :title="item.popover.title"
          @open-changed="openChanged(item, ...arguments)"
        >
          <template v-slot:trigger="{ isOpen }">
            <ToolbarButtonIcon
              class="tui-wekaToolbar__button"
              :text="item.label.toString()"
              :selected="item.active"
              :disabled="!item.enabled"
              aria-haspopup="true"
              :aria-expanded="isOpen ? 'true' : 'false'"
              @keydown.native="handleButtonKeydown"
              @focus.native="handleButtonFocus"
            >
              <template slot="icon">
                <component :is="item.iconComponent" />
              </template>
            </ToolbarButtonIcon>
          </template>
          <template v-slot:default="{ close }">
            <component :is="item.popover.component" @close="close" />
          </template>
        </Popover>
        <!-- regular toolbar button -->
        <ToolbarButtonIcon
          v-else
          :key="i"
          class="tui-wekaToolbar__button"
          :text="item.label.toString()"
          :selected="item.active"
          :disabled="!item.enabled"
          @click="itemClick(item)"
          @keydown.native="handleButtonKeydown"
          @focus.native="handleButtonFocus"
        >
          <template slot="icon">
            <component :is="item.iconComponent" />
          </template>
        </ToolbarButtonIcon>
      </template>
    </div>
  </div>
</template>

<script>
import { getTabbableElements } from 'tui/dom/focus';
import Dropdown from 'tui/components/dropdown/Dropdown';
import DropdownItem from 'tui/components/dropdown/DropdownItem';
import ShowIcon from 'tui/components/icons/Show';
import Button from 'tui/components/buttons/Button';
import ToolbarButton from 'editor_weka/components/toolbar/ToolbarButton';
import ToolbarButtonIcon from 'editor_weka/components/toolbar/ToolbarButtonIcon';
import Popover from 'tui/components/popover/Popover';

export default {
  components: {
    Dropdown,
    DropdownItem,
    ShowIcon,
    Button,
    ToolbarButton,
    ToolbarButtonIcon,
    Popover,
  },

  props: {
    items: {
      type: Array,
    },
  },

  computed: {
    activeBlockName() {
      const block = this.blockItems.find(x => x.active);
      return block ? block.label.toString() : 'Block';
    },

    blockSelectorEnabled() {
      return this.blockItems.some(x => x.enabled);
    },

    blockItems() {
      return this.items.filter(x => x.group == 'blocks');
    },

    itemGroups() {
      const items = this.items.filter(x => x.group != 'blocks');
      const groups = {
        text: [],
        embeds: [],
      };
      items.forEach(x => {
        if (!groups[x.group]) {
          groups[x.group] = [];
        }
        groups[x.group].push(x);
      });
      return groups;
    },
  },

  methods: {
    itemClick(item) {
      if (item.enabled && item.execute) {
        item.execute();
      }
    },

    openChanged(item, visible) {
      if (!visible && item.enabled && item.reset) {
        item.reset();
      }
    },

    /**
     * Handle keydown event on a button
     *
     * @param {KeyboardEvent} e
     */
    handleButtonKeydown(e) {
      const tb = e.currentTarget;

      switch (e.key) {
        case 'ArrowLeft':
        case 'Left':
          this.$_moveFocus(tb, 'prev');
          break;

        case 'ArrowRight':
        case 'Right':
          this.$_moveFocus(tb, 'next');
          break;

        case 'Home':
          this.$_moveFocus(tb, 'first');
          break;

        case 'End':
          this.$_moveFocus(tb, 'last');
          break;

        default:
          return;
      }

      e.stopPropagation();
      e.preventDefault();
    },

    /**
     * Handle focus event on a button
     *
     * @param {FocusEvent} e
     */
    handleButtonFocus(e) {
      // update tabindex
      const currentButton = e.currentTarget;
      const buttons = this.$_getAllButtons();
      buttons.forEach(button => {
        button.tabIndex = button == currentButton ? 0 : -1;
      });
    },

    /**
     * Get all toolbar buttons in order
     *
     * @returns {Element[]}
     */
    $_getAllButtons() {
      return Array.prototype.slice.call(
        this.$el.querySelectorAll('.tui-wekaToolbar__button')
      );
    },

    /**
     * Move focus from a button to a different button
     *
     * @param {Element} relativeTo
     * @param {('next'|'prev'|'first'|'last')} direction
     */
    $_moveFocus(relativeTo, direction) {
      const tabbable = getTabbableElements(this.$el);
      const buttons = this.$_getAllButtons().filter(x => tabbable.includes(x));

      const lastIndex = buttons.length - 1;

      let index = 0;
      if (direction == 'prev' || direction == 'next') {
        index = buttons.indexOf(relativeTo);
        if (index == -1) {
          index = direction == 'next' ? lastIndex : 0;
        } else {
          index += direction == 'next' ? 1 : -1;
          if (index > lastIndex) {
            index = 0;
          } else if (index < 0) {
            index = lastIndex;
          }
        }
      } else if (direction == 'first') {
        index = 0;
      } else if (direction == 'last') {
        index = lastIndex;
      }

      if (buttons[index]) {
        buttons[index].focus();
      }
    },
  },
};
</script>

<lang-strings>
{
  "editor_weka": ["format_as_blocktype_status", "label_toolbar"]
}
</lang-strings>

<style lang="scss">
.tui-wekaToolbar {
  display: flex;
  flex-wrap: wrap;
  align-items: center;
  padding: 0 var(--gap-2);
  border-bottom: 1px solid var(--color-neutral-4);

  &__group {
    display: flex;
    flex-wrap: wrap;
    align-items: center;
    margin: 0 var(--gap-2);

    > * + * {
      margin-left: 1px;
    }
  }

  &__currentBlock.tui-formBtn {
    display: flex;
    align-items: center;
    justify-content: space-between;
    min-width: 8em;
    padding: 0 var(--gap-2);

    &:disabled {
      color: var(--color-state-disabled);
    }
  }
}
</style>
