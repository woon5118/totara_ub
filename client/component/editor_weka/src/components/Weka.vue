<!--
  This file is part of Totara Enterprise Extensions.

  Copyright (C) 2020 onwards Totara Learning Solutions LTD

  Totara Enterprise Extensions is provided only to Totara
  Learning Solutions LTD's customers and partners, pursuant to
  the terms and conditions of a separate agreement with Totara
  Learning Solutions LTD or its affiliate.

  If you do not have an agreement with Totara Learning Solutions
  LTD, you may not access, use, modify, or distribute this software.
  Please contact [licensing@totaralearning.com] for more information.

  @author Simon Chester <simon.chester@totaralearning.com>
  @module editor_weka
-->

<template>
  <div :id="uid" class="tui-weka no-yui-ids">
    <Toolbar v-if="showToolbar" :items="toolbarItems" />
    <div
      ref="editorHost"
      class="tui-weka__editorHost tui-rendered"
      :data-placeholder="placeholder"
    />

    <div ref="extras">
      <!-- This is where modals/custom components will be added/removed -->
    </div>

    <div ref="extrasLive" aria-live="polite" role="status">
      <!-- This is where live-updated UI will be added/removed -->
    </div>
  </div>
</template>

<script>
import tui from 'tui/tui';
import { throttle } from 'tui/util';
import pending from 'tui/pending';
import Editor from '../js/Editor';
import Toolbar from 'editor_weka/components/toolbar/Toolbar';
import { loadLangStrings } from 'tui/i18n';
import editorWeka from 'editor_weka/graphql/weka';
import editorWekaNoSession from 'editor_weka/graphql/weka_nosession';
import FileStorage from '../js/helpers/file';
import WekaValue from '../js/WekaValue';
import { createImmutablePropWatcher } from 'tui/vue_util';

const propEqual = (val, old) => JSON.stringify(old) == JSON.stringify(val);
const warnChange = prop => createImmutablePropWatcher('Weka', prop, propEqual);

export default {
  components: {
    Toolbar,
  },

  props: {
    placeholder: {
      type: String,
      default: '',
    },
    /**
     * This property had been deprecated, please use usageIdentifier instead.
     * @deprecated since Totara 13.3
     */
    instanceId: [Number, String],
    fileItemId: [Number, String],
    /**
     * Sometimes instance id might not be set, because weka editor is used to create a new instance.
     * Therefore, this prop context id is a fallback option to replace instance's id.
     *
     * Note that this contextId will be the same as options.context_id. The reason that we don't
     * pass "context_id" via "options" was because that internally weka will look up to "options"
     * prop and decide whether to fetch from the server or not.
     */
    contextId: [Number, String],
    /**
     * This property had been deprecated, please use usageIdentifier instead.
     * @deprecated since Totara 13.3
     */
    component: String,
    /**
     * This property had been deprecated, please use usageIdentifier instead.
     * @deprecated since Totara 13.3
     */
    area: String,
    usageIdentifier: {
      type: Object,
      validator: prop => 'component' in prop && 'area' in prop,
      default() {
        // Backward compatible.
        return {
          component: this.component || 'editor_weka',
          area: this.area || 'learn',
          instanceId: this.instanceId,
        };
      },
    },
    variant: String,
    /**
     * @value {
     *   context_id: Number
     *   extensions: Object[]
     * }
     */
    options: {
      type: Object,
      validator: prop => prop.extensions !== undefined,
    },
    value: WekaValue,
    /**
     * The compact mode is to determine whether we are showing the tool bar or not.
     */
    compact: Boolean,
    /**
     * If false, loads the editor without a user logged in and disables file support.
     */
    isLoggedIn: {
      type: Boolean,
      default: true,
    },
  },

  data() {
    return {
      toolbarItems: [],

      /**
       * @deprecated since Totara 13.3
       */
      toolbarEnabled: !this.compact,
    };
  },

  computed: {
    showToolbar() {
      if (this.compact) {
        return false;
      }

      return this.toolbarItems.length > 0;
    },
  },

  watch: {
    value(value) {
      if (this.displayedValue === value) {
        return;
      }
      this.displayedValue = value || WekaValue.empty();
      if (this.editor) {
        this.editor.setValue(this.displayedValue);
      }
    },

    component: warnChange('component'),
    area: warnChange('area'),
    variant: warnChange('variant'),
    instanceId: warnChange('instanceId'),
    options: warnChange('options'),

    /**
     * @param {Number|String} value
     */
    fileItemId(value) {
      if (this.editor && this.isLoggedIn) {
        this.editor.updateFileItemId(value);
      }
    },
  },

  created() {
    this.updateToolbarThrottled = throttle(this.updateToolbar, 100);
    if (this.value) {
      this.displayedValue = this.value;
    } else {
      this.displayedValue = WekaValue.empty();
    }
    this.finalOptions = null;
  },

  mounted() {
    if (this.component) {
      console.warn(
        "The prop 'component' had been deprecated, please use 'usageIdentifier' instead"
      );
    }

    if (this.area) {
      console.warn(
        "The prop 'area' had been deprecated, please use 'usageIdentifier' instead"
      );
    }

    if (this.instanceId) {
      console.warn(
        "The prop 'instanceId' had been deprecated, please use 'usageIdentifier' instead"
      );
    }

    this.createEditor();
  },

  beforeDestroy() {
    if (this.editor) {
      this.editor.destroy();
    }
  },

  methods: {
    /**
     * Setup the editor options such as showing toolbar, the extensions that the editor needs.
     */
    async setupOptions() {
      if (this.options) {
        this.finalOptions = Object.assign({}, this.options);
        return;
      }

      // Start populating the options from the graphql call.
      const result = await this.$apollo.query({
        query: this.isLoggedIn ? editorWeka : editorWekaNoSession,
        fetchPolicy: 'no-cache',
        variables: {
          instance_id: this.usageIdentifier.instanceId,
          component: this.usageIdentifier.component,
          area: this.usageIdentifier.area,
          context_id: this.contextId || undefined,
          variant_name:
            this.variant ||
            `${this.usageIdentifier.component}-${this.usageIdentifier.area}`,
        },
      });

      this.finalOptions = Object.assign({}, result.data.editor);
    },

    /**
     * This function will try to fetch the extensions via HTTP call. This function will not cache the
     * modules that had already been loaded.
     *
     * @return {Promise<[]>}
     */
    async getExtensions() {
      await this.setupOptions();
      let extensions = [];

      if (Array.isArray(this.finalOptions.extensions)) {
        extensions = await Promise.all(
          this.finalOptions.extensions.map(({ tuicomponent, options }) => {
            let opt = {};

            if (options != null) {
              if (typeof options === 'object') {
                opt = options;
              } else {
                opt = JSON.parse(options);
              }
            }

            return tui
              .import(tuicomponent)
              .then(ext => tui.defaultExport(ext)(opt));
          })
        );
      }

      return extensions;
    },

    /**
     * This function had been deprecated and no longer used.
     *
     * @return {Promise}
     * @deprecated since Totara 13.3
     */
    async getRepositoryData() {
      console.warn(
        '[editor_weka] The function getRepositoryData had been deprecated and no longer used.'
      );

      return Promise.resolve(null);
    },

    /**
     * This function had been deprecated and no longer used.
     *
     * @return {Promise}
     * @deprecated since Totara 13.3
     */
    async getCurrentFiles() {
      console.warn(
        '[editor_weka] The function getCurrentFiles had been deprecated and no longer used.'
      );

      return Promise.resolve([]);
    },

    async createEditor() {
      if (this.editor) {
        return;
      }

      let pendingDone = pending('weka');

      const extensions = await this.getExtensions();

      let fileStorage = new FileStorage({
        itemId: this.fileItemId,
        contextId: this.finalOptions.context_id || null,
      });

      this.editor = new Editor({
        value: this.displayedValue,
        placeholder: this.placeholder,
        parent: this,
        viewExtrasEl: this.$refs.extras,
        viewExtrasLiveEl: this.$refs.extrasLive,
        extensions: extensions,
        fileStorage: fileStorage,
        onUpdate: this.$_onUpdate.bind(this),
        contextId: this.finalOptions.context_id || null,
        component: this.usageIdentifier.component || null,
        area: this.usageIdentifier.area || null,
        instanceId: this.usageIdentifier.instanceId || null,
        onTransaction: () => {
          this.updateToolbarThrottled();
        },
        onFocus: e => {
          this.$emit('focus', e);
        },
        onBlur: e => {
          this.$emit('blur', e);
        },
      });

      const components = this.editor.allVueComponents();

      await Promise.all([
        tui.loadRequirements({ components }),
        loadLangStrings(this.editor.allStrings()),
      ]);

      this.view = this.editor.createView(this.$refs.editorHost);

      this.updateToolbar();

      pendingDone();

      // Event emitted to make the parent component knowing that this editor has been mounted properly.
      this.$emit('ready');
    },

    updateToolbar() {
      if (!this.editor.destroyed) {
        this.toolbarItems = this.editor.getToolbarItems();
      }
    },

    /**
     * Emit new value.
     *
     * @param {WekaValue} value
     */
    $_onUpdate(value) {
      this.displayedValue = value;
      this.$emit('input', value);
    },
  },
};
</script>

<style lang="scss">
.tui-weka {
  position: relative;
  display: flex;
  flex-direction: column;
  width: 100%;
  background-color: var(--color-neutral-1);
  border: var(--border-width-thin) solid var(--form-input-border-color);

  &__placeholder {
    // Styling for the place holder.
    &:before {
      color: var(--color-neutral-6);
      content: attr(data-placeholder);
    }
  }

  &__editorHost {
    display: flex;
    flex-direction: column;
    flex-grow: 1;
    overflow: auto;

    > .tui-weka-editor {
      flex-grow: 1;
    }
  }

  .ProseMirror-focused {
    .tui-weka {
      &__placeholder {
        &:before {
          content: '';
        }
      }
    }
  }

  .ProseMirror {
    padding: var(--gap-4);
    white-space: pre-wrap;
    white-space: break-spaces;
    word-wrap: break-word;
    -webkit-font-variant-ligatures: none;
    font-variant-ligatures: none;
    font-feature-settings: 'liga' 0; /* the above doesn't seem to work in Edge */

    &:focus {
      outline: none;
    }

    hr {
      margin: 0 0 var(--gap-2) 0;
    }

    pre {
      white-space: pre-wrap;
    }

    ol,
    ul {
      margin: 0;
      padding-left: var(--gap-4);
    }

    ul ul {
      list-style-type: circle;
    }
    ul ul ul {
      list-style-type: square;
    }

    li {
      position: relative;
    }
  }

  .ProseMirror-hideselection *::selection,
  .ProseMirror-hideselection *::-moz-selection {
    background: transparent;
  }

  .ProseMirror-hideselection {
    caret-color: transparent;
  }

  .ProseMirror-selectednode {
    outline: var(--border-width-normal) solid var(--color-secondary);
  }

  /* Make sure li selections wrap around markers */

  li.ProseMirror-selectednode {
    outline: none;
  }

  li.ProseMirror-selectednode:after {
    position: absolute;
    top: -2px;
    right: -2px;
    bottom: -2px;
    left: -32px;
    border: var(--border-width-normal) solid var(--color-secondary);
    content: '';
    pointer-events: none;
  }

  .ProseMirror-gapcursor:before {
    // insert an nbsp to make gapcursor expand to full line height
    content: '\00a0';
  }
}
</style>
