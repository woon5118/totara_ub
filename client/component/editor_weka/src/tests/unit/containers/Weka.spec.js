/**
 * This file is part of Totara Enterprise Extensions.
 *
 * Copyright (C) 2020 onwards Totara Learning Solutions LTD
 *
 * Totara Enterprise Extensions is provided only to Totara
 * Learning Solutions LTD's customers and partners, pursuant to
 * the terms and conditions of a separate agreement with Totara
 * Learning Solutions LTD or its affiliate.
 *
 * If you do not have an agreement with Totara Learning Solutions
 * LTD, you may not access, use, modify, or distribute this software.
 * Please contact [licensing@totaralearning.com] for more information.
 *
 * @author Alvin Smith <alvin.smith@totaralearning.com>
 * @module editor_weka
 */

import { shallowMount } from '@vue/test-utils';
import Toolbar from 'editor_weka/components/toolbar/Toolbar';
import Weka from 'editor_weka/components/Weka';
import Vue from 'vue';

// Mock tui.import
jest.mock('tui/tui', function() {
  return {
    import: async function(id) {
      switch (id) {
        case 'editor_weka/extensions/link':
          return require('../../../js/extensions/link');

        case 'editor_weka/extensions/text':
          return require('../../../js/extensions/text');

        default:
          throw `No module found for id ${id}`;
      }
    },

    defaultExport(x) {
      return x.default || x;
    },

    loadRequirements: () => new Promise(resolve => resolve()),
  };
});

jest.mock('editor_weka/api', () => ({
  __esModule: true,
  getLinkMetadata: () => null,
}));

const EXTENSIONS = [
  {
    name: 'link',
    tuicomponent: 'editor_weka/extensions/link',
  },
  {
    name: 'text',
    tuicomponent: 'editor_weka/extensions/text',
  },
];

/**
 *
 * @param {Object} option
 * @param {Number} instanceId
 * @param {Boolean} compact
 * @return {Wrapper<Vue>}
 */
const factory = (option, instanceId, compact) => {
  return shallowMount(Weka, {
    propsData: {
      options: option,
      usageIdentifier: {
        component: 'editor_weka',
        area: 'default',
        instanceId: instanceId,
      },
      compact: compact,
      ariaLabel: 'Weka label',
    },
    mocks: {
      $apollo: {
        query: query => {
          if (query.variables.instance_id) {
            return {
              data: { editor: { showtoolbar: true, extensions: EXTENSIONS } },
            };
          }
        },
      },
      uid: 'uid-weka',
    },
  });
};

describe('editor_weka/components/Weka.vue', () => {
  it('toolbar is showing and render correctly', async () => {
    const wrapper = factory({ extensions: EXTENSIONS }, 42, false);

    // Since the editor would have to be mounted once all the elements within the component rendered, therefore,
    // this test should be waiting for that mounting event to be finished to run the assertion.
    await new Promise(resolve => {
      wrapper.vm.$on('ready', () => {
        resolve('mounted');
      });
    });

    expect(wrapper.find(Toolbar).exists()).toBeTrue();
    expect(wrapper.element).toMatchSnapshot();
  });

  it('toolbar is hidden when showtoolbar is set to false no matter extension load or not', async () => {
    const wrapper = factory({ extensions: EXTENSIONS }, null, true);
    wrapper.setData({ toolbarItems: [{ foo: 'bar' }] });
    await Vue.nextTick();
    expect(wrapper.find(Toolbar).exists()).toBeFalse();
  });

  it('toolbar is showing when showtoolbar is not set and there are extensions', async () => {
    const wrapper = factory(null, 15);
    wrapper.setData({ toolbarItems: [{ foo: 'bar' }] });
    await Vue.nextTick();
    expect(wrapper.find(Toolbar).exists()).toBeTrue();
  });
});
