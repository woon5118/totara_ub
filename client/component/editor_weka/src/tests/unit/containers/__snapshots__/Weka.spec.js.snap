// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`editor_weka/components/Weka.vue toolbar is showing and render correctly 1`] = `
<div
  arialabel="Weka label"
  class="tui-weka no-yui-ids"
  id="uid-weka"
>
  <toolbar-stub
    items="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]"
  />
   
  <div
    class="tui-weka__editorHost tui-rendered"
    data-placeholder=""
  >
    <div
      aria-label="[[pluginname, editor_weka]]"
      class="ProseMirror tui-weka-editor"
      contenteditable="true"
      role="region"
    >
      <p>
        <br />
      </p>
    </div>
  </div>
   
  <div />
   
  <div
    aria-live="polite"
    role="status"
  />
</div>
`;
